### PR TITLE
Convert ModelPT to optionally load datasets

### PR DESCRIPTION
**Pull request by titu1994**
# Bugfix
- ModelPT tries to load train, val and test data loaders without safety checks.
- This causes an issue upon deserialization via ModelPT.restore_from(), which holds old paths to manifest/datasets etc which are invalid on a new system.
- This patch optionally attempts to instantiate the data loaders, which failing and logging a warning with guidance to user to prepare data loaders explicitly via some methods.

# Caveats
- Due to the automatic multi data loader support for validation and test ds, and since subclasses override `setup_validation_data` and `setup_test_data`, it is not possibly to re-perform the multi-dataloader logic by using just the setup* calls.
- To get around this, there are 2 ways.
  - Use another method (here called `prepare_*` methods with same signature to perform maintenance work
  - Use a meta class to wrap all subclass overrides of `setup_validation_data` and `setup_test_data` with this maintenance work. In doing so, user can continue calling `setup_*` methods and be unaware that we are doing multi loader maintenance.
- This PR takes approach 1 to avoid over complicated book keeping, but approach 2 is more stream lined and in principle reduces burdon on users. 

Signed-off-by: smajumdar <titu1994@gmail.com>

### CODE REVIEW

**Comment by titu1994** [comment]
This is a grest suggestion. Setup validation and test are abstract so they must be implemented in some form (even empty implementations). PyTorch lightning has a method to check if a subclass implements a method, but we can't check if they are empty implementations or not. What we can do is after calling the method, we can inspection the existence of a non none val and test data loaders ANS warn users about it.

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -80,32 +80,23 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
-                # Set some placeholder overriden by helper method
-                self._validation_loss_idx = 0
-                self._validation_names = None
-                self._validation_dl = None  # type: torch.utils.data.DataLoader
+            try:
+                if 'train_ds' in self._cfg and self._cfg.train_ds is not None:
+                    self.setup_training_data(self._cfg.train_ds)
 
-                model_utils.resolve_validation_dataloaders(model=self)
+            except Exception:
```
Comment (blisc) [inline nemo/core/classes/modelPT.py:87]: Can we stop catching `Exceptions`? Catch the `Exception` you expect to occur, otherwise just let it fail.
If you want to print a helper message, then feel free to print the helper function, but then reraise the old Exception.

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -80,32 +80,23 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self._trainer = trainer
```
Comment (blisc) [inline nemo/core/classes/modelPT.py:81]: Also a more fundamental question, why does restore_from() attempt to restore dataloaders?

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:87]: There are upto 7 (as far as far as I know) messages that can occur here, cause this is dependent on several subclass code behaviour, modalities of operation - single vs multiple data loaders, different ways of execution - fadtvpath vs explicit path resolution, different domains handling the Dataloaders differently. We also do not want to raise the exception as then we cannot resume model loading. 

At worst, I can look into catching the exception stack trace and logging it into a debug level, but catching explicitly each exception is not reasonable here.

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:81]: Restore from calls model constrictor, which in turn loads the new or old cfg file, which in turn has the current or old path to the data directory, which in turn crashes resume if run on a new machine. This is a safety net against that. Seems @nithinraok has already faced issues cause of this behaviour.

Comment (blisc) [reply nemo/core/classes/modelPT.py:81]: So why do we do this?
Why not just remove all dataloaders from config when restoring?
Why not have a restore_weights_only function?
If the design has flaws, we can change it, instead of hacking around it.

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:87]: I'm fact without this patch, restore from .Nemo files is basically impossible on different systems than the one that was used to generate the .Nemo file.

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (nithinraok) [reply nemo/core/classes/modelPT.py:81]: yes, my suggestion is to have a separate section in cfg to mention data_related stuff and have a setup_data() method which calls  set_up_{test, train, valid}_data() methods to load dataloaders. 
This way when we save or restore model only save model related stuff and not interfere with data.

Comment (titu1994) [reply nemo/core/classes/modelPT.py:81]: @okuchaiev Please add points here. 
From my view, suggestion to remove dataloader from serialization means the user has to write their own config after installing model from a .Nemo file cthis defeats the purpose of .Nemo. Instead, the proposed workflow which @okuchaiev is working on is -
1) Load the checkpoint, bypass the incorrect config and show it to the user in the log. 
2) User overwrites the critical path directly in Python (OmegaConf)
3) them call the appropriate setup function which this data loader. We overwrite the self._cfg.*_ds definition and use the override just as always. 

This allows the user 
1) to have a predefined config with almost every value already set
2) to have something that can be overriden in Python rather than mess around with more yaml
3) to safely support packaging all the remaining items - tokemixers, and data files from NLP also fall under this. Not all domains follow the manifest approach, but this allos override by user at any level since they have access to the full cfg after instantiation (well everything other than the trainer config)

Comment (titu1994) [reply nemo/core/classes/modelPT.py:81]: @nithinraok As we discussed in the call, thus is not a useful solution. NLP, CV don't follow the same rules and structure that ASR, does not does TTS. You are thinking only from the view of ASR. The data loader config would again have to be defined by the user for your solution, entirely defeating the purpose of the packaged .Nemo file

Comment (titu1994) [reply nemo/core/classes/modelPT.py:81]: As to having a restore_weigits_only, again, Core cannot make assumptions about which parts of the manifest needs paths. Look at NLP manifests, their structure is entirely different than ASR. Look at CV which doesn't even have paths. Why should core delete configs which declare splits of train eval test for a dataset and force the user to redefine it in Yaml or python ?

Comment (blisc) [reply nemo/core/classes/modelPT.py:81]: > Load the checkpoint, bypass the incorrect config and show it to the user in the log.

So it seems to me, the answer to should restore_from() attempt to restore dataloaders is: No, restore_from() should not attempt to restore dataloaders.
Isn't that what you are doing? Remove the offending dataloader parts from the config, instantiate the model, and then instantiate the datasets.

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (blisc) [reply nemo/core/classes/modelPT.py:87]: Ok, let's simply just print the original exception as well and then we can merge.

File: nemo/core/classes/modelPT.py [hunk 3]
```diff
@@ -239,6 +230,72 @@ def setup_test_data(self, test_data_config: Union[DictConfig, Dict]):
...
+
+                self._cfg.validation_ds = val_data_config
+
+            model_utils.resolve_validation_dataloaders(model=self)
+
+            if self._validation_names is None:
+                if self._validation_dl is not None and type(self._validation_dl) in [list, tuple]:
+                    self._validation_names = ['val_{}_'.format(idx) for idx in range(len(self._validation_dl))]
+
+        except Exception:
```
Comment (blisc) [inline nemo/core/classes/modelPT.py:258]: Please print the Exception here as well.

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -239,6 +230,72 @@ def setup_test_data(self, test_data_config: Union[DictConfig, Dict]):
...
+
+                self._cfg.test_ds = test_data_config
+
+            model_utils.resolve_test_dataloaders(model=self)
+
+            if self._test_names is None:
+                if self._test_dl is not None and type(self._test_dl) in [list, tuple]:
+                    self._test_names = ['test_{}_'.format(idx) for idx in range(len(self._test_dl))]
+
+        except Exception:
```
Comment (blisc) [inline nemo/core/classes/modelPT.py:291]: Please print the Exception here as well.

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:81]: Yes, in a sense. What I try is case 1) original model owner is training so path is correct (if not, I will print the warning with full stack trace under debug level). They then export .Nemo. 

Case 2) new user on another machine uses .Nemo file to restore in say a Jupyter notebook or script. This raises the warning (but user likely will not use debug level logging so no stack trace shown). 

They get an example of the entire validation of test config that was attempted to load. They can then deepcopy the mldel._cfg and then update model._cfg.*_ds.* to override whatever offending path (or data or tokenizer or split etc). Finally they follow the instructions in the warning and call the appropriate setup method with this updated config, thereby successfully loading their own data + whatever other config changes to the dataset they want.

This way, yes, we safely notify the user that the manifest was incorrect, without making assumptions as to how the manifest is built - this allows cross domain flexibility. 

We show the user the config, so they can edit as they please. This is lost if we destroy the config themselves during save. Restore from also can't make that assumptions as to how data loading will occur across domains. So this simply sidesteps the issue across domains.

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py]: Done

File: nemo/core/classes/modelPT.py (hunk 3 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py]: Done

File: nemo/core/classes/modelPT.py (hunk 4 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py]: Done

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (blisc) [reply nemo/core/classes/modelPT.py]: I might even think we should move it up from debug to warning, so that user's can see the error to have an indication what in their config they should fix.

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (okuchaiev) [reply nemo/core/classes/modelPT.py:82]: @blisc  The promise behind the model is that it contains all the necessary details for training/validation/testing.
So "Why not just remove all dataloaders from config when restoring"  - because there are important information in the dataloader config. For example, for ASR models - sample_rate, batch_size, normalize transcripts or not, for BERT whether max token was 128 or 512, etc.

If would be nice if we could keep those details and drop paths to local files/folders. But, unfortunately, those are handled differently from model to model. Hence, this patch.
